In [2]:
import main

INFO:tensorflow:Enabling eager execution
INFO:tensorflow:Enabling v2 tensorshape
INFO:tensorflow:Enabling resource variables
INFO:tensorflow:Enabling tensor equality
INFO:tensorflow:Enabling control flow v2


In [2]:
b = main.Board()

In [3]:
import numpy as np

In [4]:
np.array([1,2,3])

array([1, 2, 3])

In [5]:
np.array(["hi"])

array(['hi'], dtype='<U2')

In [6]:
np.array([1,2,3]) == np.array([1,1,1])

array([ True, False, False])

In [7]:
for x in range(10):
    print(x)

0
1
2
3
4
5
6
7
8
9


In [9]:
for x in range(0, 10,  -1):
    print(x)

In [10]:
range?

In [12]:
for x in range(10, 0,  -1):
    print(x)

10
9
8
7
6
5
4
3
2
1


In [16]:
list(range(10)).reverse()

In [1]:
import main

INFO:tensorflow:Enabling eager execution
INFO:tensorflow:Enabling v2 tensorshape
INFO:tensorflow:Enabling resource variables
INFO:tensorflow:Enabling tensor equality
INFO:tensorflow:Enabling control flow v2


In [ ]:
main.run_game()

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

Move: H11
['H11']
Duration of game so far is 6.637568950653076 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++

Move: H13
['H11', 'H12', 'H13']
Duration of game so far is 9.495049238204956 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●○○○++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Right
Moves made: 3
Ball at: [7 9]

[[[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  

Move: H16
['H11', 'H12', 'H13', 'H14', 'H15', 'H16']
Duration of game so far is 12.871709108352661 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●○○○○○○+++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 6
Ball at: [7 9]

[[[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
 

 import main

In [1]:
import main

INFO:tensorflow:Enabling eager execution
INFO:tensorflow:Enabling v2 tensorshape
INFO:tensorflow:Enabling resource variables
INFO:tensorflow:Enabling tensor equality
INFO:tensorflow:Enabling control flow v2


In [2]:
main.run_game(players = [main.PloddingPlayer(),main.NegamaxABPlayer(static_evaluator=main.CNNEvaluator(),)])

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

['H11']
Duration of game so far is 0.00494384765625 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●○++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Right
Moves made: 1
Ball at: [7 9]

Applying static evaluator to current position:
0.322417676448822
Initial score is 0
New best move is F8 which has score of 0.28

New best move is F9 which has score of 0.21674352884292603
New best move is G9 which has score of 0.3430069088935852
New best move is H9 which has score of 0.389950156211853
53 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9']
Duration of game so far is 9.69101095199585 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H++++++○+○●○+○++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 12
Ball at: [7 9]

['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ']
Duration of game so far is 9.698030948638916 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++

New best move is H9 which has score of 0.6098082065582275
67 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9']
Duration of game so far is 19.008694171905518 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H++++++○+○●++++○++++
I+++++++++○+++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 22
Ball at: [7 9]

['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11']
Duration of game so far is 19.014893770217896 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C++++++++++++++++

New best move is E8 which has score of 0.43291425704956055
New best move is F11 which has score of 0.45585066080093384
New best move is G10 which has score of 0.4586435556411743
New best move is H9 which has score of 0.484994113445282
New best move is I9 which has score of 0.5134049654006958
67 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9']
Duration of game so far is 29.665520906448364 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G++++++++○++++++++++
H++++++○++●○+++○++++
I++++++++○++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 32
Ball at: [7 9]

['H11'

New best move is F7 which has score of 0.36078065633773804
New best move is F10 which has score of 0.36559760570526123
New best move is G7 which has score of 0.36931276321411133
New best move is G10 which has score of 0.3796049952507019
New best move is I9 which has score of 0.5008597373962402
New best move is J9 which has score of 0.5737487077713013
79 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9']
Duration of game so far is 41.46477484703064 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G++++++++○++++++++++
H++++++○+++○+○+○++++
I+++++++++++++++++++
J++++++○+○●+++++++++
K+++++++++++++++++++
L++++++++++++++++

New best move is F7 which has score of 0.29423773288726807
New best move is F8 which has score of 0.306755006313324
New best move is H9 which has score of 0.30725234746932983
New best move is I7 which has score of 0.3525465130805969
New best move is I8 which has score of 0.39377111196517944
New best move is K7 which has score of 0.4009494185447693
94 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7']
Duration of game so far is 52.0357871055603 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G++++++++○++++++++++
H++++++○+++○+○+○++++
I+++++++++++++++++++
J+++++++●○+○++

New best move is F7 which has score of 0.6103181838989258
New best move is G7 which has score of 0.6129806041717529
New best move is I6 which has score of 0.6147567629814148
New best move is J6 which has score of 0.617851972579956
New best move is J7 which has score of 0.6310417652130127
New best move is K6 which has score of 0.6556375026702881
New best move is M5 which has score of 0.6817383766174316
104 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5']
Duration of game so far is 64.4005708694458 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++

New best move is G7 which has score of 0.6504884362220764
New best move is K4 which has score of 0.6782277226448059
New best move is M3 which has score of 0.7375174760818481
117 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5', 'E ', 'L7', 'L9', 'W ', 'L7', 'SW ', 'N5', 'M3']
Duration of game so far is 78.39473485946655 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G++++++++○++++++++++
H++++++○+++○+○+○++++
I+++++++++++++++++++
J++++++++++++○++++++
K++++○++++++++++++++
L++++++○+○++++++++++
M++○++++++++++++++++


New best move is F7 which has score of 0.480594277381897
New best move is G7 which has score of 0.4839571714401245
New best move is J7 which has score of 0.4868413805961609
New best move is K6 which has score of 0.533204972743988
New best move is M5 which has score of 0.5388028025627136
New best move is N5 which has score of 0.5730322599411011
New best move is E  which has score of 0.5769324898719788
New best move is E NW NW  which has score of 0.617275595664978
New best move is NE W  which has score of 0.6942162215709686
131 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5', 'E ', 'L7', 'L9', 'W ', 'L7', 'SW ', 'N5', 'M3', 'E ', 'N5', 'N

New best move is F7 which has score of 0.6078410148620605
New best move is F9 which has score of 0.6080337762832642
New best move is G7 which has score of 0.630609542131424
New best move is W  which has score of 0.8808107227087021
132 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5', 'E ', 'L7', 'L9', 'W ', 'L7', 'SW ', 'N5', 'M3', 'E ', 'N5', 'N7', 'W ', 'N5', 'M7', 'E ', 'NE W ', 'L7', 'NW ', 'J5', 'J3', 'E ', 'J5', 'J7', 'W ']
Duration of game so far is 109.81605315208435 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++

New best move is F7 which has score of 0.5006375908851624
New best move is F9 which has score of 0.5061248540878296
New best move is G6 which has score of 0.5131720304489136
New best move is I5 which has score of 0.5832076072692871
New best move is J5 which has score of 0.5869038701057434
New best move is K5 which has score of 0.6136559844017029
133 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5', 'E ', 'L7', 'L9', 'W ', 'L7', 'SW ', 'N5', 'M3', 'E ', 'N5', 'N7', 'W ', 'N5', 'M7', 'E ', 'NE W ', 'L7', 'NW ', 'J5', 'J3', 'E ', 'J5', 'J7', 'W ', 'J5', 'W ', 'J3', 'I1', 'E ', 'J3', 'E ', 'K5']
Duration of game so far is 125.83840107917786 

New best move is J4 which has score of 0.6357756853103638
New best move is K4 which has score of 0.667951375246048
New best move is K7 which has score of 0.6719220280647278
New best move is SW  which has score of 0.7501508891582489
135 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5', 'E ', 'L7', 'L9', 'W ', 'L7', 'SW ', 'N5', 'M3', 'E ', 'N5', 'N7', 'W ', 'N5', 'M7', 'E ', 'NE W ', 'L7', 'NW ', 'J5', 'J3', 'E ', 'J5', 'J7', 'W ', 'J5', 'W ', 'J3', 'I1', 'E ', 'J3', 'E ', 'K5', 'E ', 'J7', 'J9', 'W SW ', 'L5', 'SW ']
Duration of game so far is 137.8635768890381 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++

New best move is K2 which has score of 0.7144149541854858
New best move is K4 which has score of 0.7362688779830933
New best move is K5 which has score of 0.73959881067276
New best move is K7 which has score of 0.7454894781112671
New best move is M7 which has score of 0.759421095252037
New best move is NW  which has score of 0.8053236305713654
137 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5', 'E ', 'L7', 'L9', 'W ', 'L7', 'SW ', 'N5', 'M3', 'E ', 'N5', 'N7', 'W ', 'N5', 'M7', 'E ', 'NE W ', 'L7', 'NW ', 'J5', 'J3', 'E ', 'J5', 'J7', 'W ', 'J5', 'W ', 'J3', 'I1', 'E ', 'J3', 'E ', 'K5', 'E ', 'J7', 'J9', 'W SW ', 'L5', 'SW ', 'N3', 'M

New best move is F7 which has score of 0.46075165271759033
New best move is H9 which has score of 0.46997034549713135
New best move is I7 which has score of 0.47019416093826294
New best move is I8 which has score of 0.4897449016571045
New best move is I9 which has score of 0.4975787401199341
New best move is K5 which has score of 0.5518563389778137
New best move is K9 which has score of 0.5657027959823608
130 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5', 'E ', 'L7', 'L9', 'W ', 'L7', 'SW ', 'N5', 'M3', 'E ', 'N5', 'N7', 'W ', 'N5', 'M7', 'E ', 'NE W ', 'L7', 'NW ', 'J5', 'J3', 'E ', 'J5', 'J7', 'W ', 'J5', 'W ', 'J3', 'I1', 'E ', 'J3

New best move is F7 which has score of 0.3095729351043701
New best move is F8 which has score of 0.322090208530426
New best move is I7 which has score of 0.36330050230026245
New best move is I8 which has score of 0.4077240228652954
New best move is K7 which has score of 0.4231565594673157
131 calls to static evaluator
['H11', 'H9', 'E ', 'H11', 'H13', 'W ', 'H11', 'W ', 'H9', 'H7', 'E ', 'H9', 'E ', 'H11', 'E ', 'H13', 'H15', 'W W W ', 'H9', 'I10', 'E ', 'H9', 'H11', 'W ', 'H9', 'I11', 'E ', 'G9', 'E ', 'SW N ', 'H11', 'I9', 'E ', 'H11', 'H13', 'W ', 'H11', 'SW ', 'J9', 'J7', 'E ', 'J9', 'J11', 'W ', 'J9', 'W ', 'J7', 'K5', 'E ', 'K7', 'E ', 'J9', 'E ', 'J11', 'J13', 'W W SW ', 'L7', 'M5', 'E ', 'L7', 'L9', 'W ', 'L7', 'SW ', 'N5', 'M3', 'E ', 'N5', 'N7', 'W ', 'N5', 'M7', 'E ', 'NE W ', 'L7', 'NW ', 'J5', 'J3', 'E ', 'J5', 'J7', 'W ', 'J5', 'W ', 'J3', 'I1', 'E ', 'J3', 'E ', 'K5', 'E ', 'J7', 'J9', 'W SW ', 'L5', 'SW ', 'N3', 'M4', 'E ', 'M3', 'N5', 'NW ', 'L3', 'L1', 'E ', 'K4', 'E 

New best move is F7 which has score of 0.19877856969833374
New best move is G7 which has score of 0.19909679889678955
New best move is G10 which has score of 0.2067970633506775
New best move is I13 which has score of 0.2755891680717468
New best move is J13 which has score of 0.409986674785614


KeyboardInterrupt: 

In [3]:
main.run_game(players = [main.PloddingPlayer(),main.NegamaxABPlayer(depth = 2, static_evaluator=main.CNNEvaluator(),)])

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

['H11']
Duration of game so far is 0.0053539276123046875 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●○++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Right
Moves made: 1
Ball at: [7 9]

Applying static evaluator to current position:
0.322417676448822
Initial score is 0
New best move is F8 which has score of

New best move is E9 which has score of 0.23615361750125885
New best move is E10 which has score of 0.24165472388267517
New best move is E11 which has score of 0.2472575306892395
New best move is E13 which has score of 0.2756882309913635
New best move is G13 which has score of 0.3494585156440735
New best move is I13 which has score of 0.353647381067276
New best move is NW  which has score of 0.4580368399620056
4425 calls to static evaluator
['H11', 'H9', 'E ', 'I11', 'H13', 'G11', 'E ', 'H13', 'H15', 'W ', 'H13', 'NW ']
Duration of game so far is 482.1036608219147 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++●+++++++++
G+++++++++++++++++++
H++++++++○+++○+○++++
I++++++++++○++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 12
Ball at: [5 9]

['H11', 'H9', '

New best move is C7 which has score of 0.35309645533561707
New best move is C8 which has score of 0.3563389182090759
New best move is C11 which has score of 0.38307684659957886
New best move is E11 which has score of 0.4605015516281128
New best move is NW  which has score of 0.6019827723503113
8085 calls to static evaluator
['H11', 'H9', 'E ', 'I11', 'H13', 'G11', 'E ', 'H13', 'H15', 'W ', 'H13', 'NW ', 'F11', 'E9', 'E ', 'E11', 'F13', 'NW SW ', 'F9', 'F7', 'E ', 'E9', 'F11', 'NW ']
Duration of game so far is 1485.8404870033264 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++●+++++++++++
E+++++++++++++++++++
F++++++○+++○+○++++++
G+++++++++++++++++++
H++++++++○+++○+○++++
I++++++++++○++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 24
Ball at: [3 7]

['H11', 'H9', 'E ', 'I11', 'H13', 'G11', 'E ', 'H13

New best move is A4 which has score of 0.6602791547775269
New best move is A5 which has score of 0.690479040145874
New best move is B5 which has score of 0.7398813366889954
New best move is C5 which has score of 0.7577927708625793
11431 calls to static evaluator
['H11', 'H9', 'E ', 'I11', 'H13', 'G11', 'E ', 'H13', 'H15', 'W ', 'H13', 'NW ', 'F11', 'E9', 'E ', 'E11', 'F13', 'NW SW ', 'F9', 'F7', 'E ', 'E9', 'F11', 'NW ', 'D9', 'C7', 'E ', 'D9', 'D11', 'W ', 'D9', 'NW ', 'B7', 'C5']
Duration of game so far is 2807.37553691864 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++●○++++++++++++
C++++○++++++++++++++
D++++++++○+○++++++++
E+++++++++++++++++++
F++++++○+++○+○++++++
G+++++++++++++++++++
H++++++++○+++○+○++++
I++++++++++○++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 34
Ball at: [1 5]

['H11', 'H9', 'E ', 'I11', 'H13', 'G11', 'E ', 'H13', 

New best move is A4 which has score of 0.6681755781173706
New best move is A5 which has score of 0.689638614654541
New best move is B5 which has score of 0.8120916485786438
New best move is C5 which has score of 0.8853850960731506
14749 calls to static evaluator
['H11', 'H9', 'E ', 'I11', 'H13', 'G11', 'E ', 'H13', 'H15', 'W ', 'H13', 'NW ', 'F11', 'E9', 'E ', 'E11', 'F13', 'NW SW ', 'F9', 'F7', 'E ', 'E9', 'F11', 'NW ', 'D9', 'C7', 'E ', 'D9', 'D11', 'W ', 'D9', 'NW ', 'B7', 'C5', 'E ', 'C7', 'B9', 'SW ', 'D7', 'NW ', 'B5', 'C3', 'E ', 'C5']
Duration of game so far is 4486.18231511116 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++●++○++++++++++
C++○+○++++++++++++++
D++++++○+○+○++++++++
E+++++++++++++++++++
F++++++○+++○+○++++++
G+++++++++++++++++++
H++++++++○+++○+○++++
I++++++++++○++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 44
Ball at:

Yes, Plodding Player has been beaten! Right is taking about 228 sec per move (5940/26), and making 15000 or less calls to the static evaluator per move.

In [4]:
main.run_game(players = [main.NegamaxABPlayer(depth = 2, static_evaluator=main.LocationEvaluator(),),main.NegamaxABPlayer(depth = 2, static_evaluator=main.CNNEvaluator(),)])

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

Applying static evaluator to current position:
0.5
Initial score is 0
New best move is F8 which has score of 0.5
New best move is F9 which has score of 0.5
New best move is F10 which has score of 0.5
New best move is F11 which has score of 0.5
New best move is F12 which has score of 0.5
New best move is G8 which has score of 0.5
New best move is G10 which has score of 0.5
New best move is G11 which has score of 0.5
New best move is G12 which has score of 0.5
New best move is H8 which has score of 0.5
New best move is H11 which has score of 0.5
New best move is H12 which has score of 0.5


0.8531647324562073
Initial score is 0
New best move is D4 which has score of 0.6588152647018433
New best move is D5 which has score of 0.6606299877166748
New best move is D6 which has score of 0.6638505458831787
New best move is D7 which has score of 0.6928827166557312
New best move is E5 which has score of 0.818507730960846
New best move is F5 which has score of 0.8384941220283508
2780 calls to static evaluator
['J12', 'G9', 'NW ', 'F7', 'W ', 'F5']
Duration of game so far is 203.52355313301086 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F++++○●+++++++++++++
G+++++++++++++++++++
H+++++++++++++++++++
I+++++++++++++++++++
J+++++++++++○+++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 6
Ball at: [5 5]

Applying static evaluator to current position:
0.2777777777777778
Initial score is 0


New best move is B1 which has score of 0.9477152824401855
New best move is B3 which has score of 0.9769508838653564
New best move is C1 which has score of 0.9834926128387451
New best move is E1 which has score of 0.9851722717285156
2217 calls to static evaluator
['J12', 'G9', 'NW ', 'F7', 'W ', 'F5', 'W ', 'E3', 'NW ', 'E1']
Duration of game so far is 341.15141010284424 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+●+++++++++++++++++
E○++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++++++++++++
I+++++++++++++++++++
J+++++++++++○+++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 10
Ball at: [3 1]

Applying static evaluator to current position:
0.05555555555555555
Initial score is 0
New best move is B1 which has score of 0.0
New best move is B2 which has score of 0.0
New best move is B3 which has score of 0

2-ply with location evaluator loses badly to 2-ply with CNN. The former can't help playing a move that'll help its opponent. 

But the former also moves in a fraction of a second, whereas the latter takes roughly a minute per move. 
And they're both using the static evaluator a few thousand times. So this calls for profiling.

In [5]:
import cProfile

In [7]:
cProfile.run("main.run_game(players = [main.NegamaxABPlayer(depth = 2, static_evaluator=main.LocationEvaluator(),),main.NegamaxABPlayer(depth = 2, static_evaluator=main.CNNEvaluator(),)])")

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

Applying static evaluator to current position:
0.5
Initial score is 0
New best move is F8 which has score of 0.5
New best move is F9 which has score of 0.5
New best move is F10 which has score of 0.5
New best move is F11 which has score of 0.5
New best move is F12 which has score of 0.5
New best move is G8 which has score of 0.5
New best move is G10 which has score of 0.5
New best move is G11 which has score of 0.5
New best move is G12 which has score of 0.5
New best move is H8 which has score of 0.5
New best move is H11 which has score of 0.5
New best move is H12 which has score of 0.5


New best move is D4 which has score of 0.6588152647018433
New best move is D5 which has score of 0.6606299877166748
New best move is D6 which has score of 0.6638505458831787
New best move is D7 which has score of 0.6928827166557312
New best move is E5 which has score of 0.818507730960846
New best move is F5 which has score of 0.8384941220283508
2780 calls to static evaluator
['J12', 'G9', 'NW ', 'F7', 'W ', 'F5']
Duration of game so far is 235.44834113121033 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F++++○●+++++++++++++
G+++++++++++++++++++
H+++++++++++++++++++
I+++++++++++++++++++
J+++++++++++○+++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 6
Ball at: [5 5]

Applying static evaluator to current position:
0.2777777777777778
Initial score is 0
New best move is D3 which has score of

New best move is B1 which has score of 0.9477152824401855
New best move is B3 which has score of 0.9769508838653564
New best move is C1 which has score of 0.9834926128387451
New best move is E1 which has score of 0.9851722717285156
2217 calls to static evaluator
['J12', 'G9', 'NW ', 'F7', 'W ', 'F5', 'W ', 'E3', 'NW ', 'E1']
Duration of game so far is 398.74573802948 seconds
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+●+++++++++++++++++
E○++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++++++++++++
I+++++++++++++++++++
J+++++++++++○+++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 10
Ball at: [3 1]

Applying static evaluator to current position:
0.05555555555555555
Initial score is 0
New best move is B1 which has score of 0.0
New best move is B2 which has score of 0.0
New best move is B3 which has score of 0.0


    36740    0.018    0.000    0.023    0.000 context.py:1875(global_seed)
  1176964    0.884    0.000    2.396    0.000 context.py:1885(executing_eagerly)
       40    0.000    0.000    0.000    0.000 context.py:2014(shared_name)
   208326    0.167    0.000    0.443    0.000 context.py:2035(graph_mode)
   147030    0.112    0.000    0.367    0.000 context.py:2041(eager_mode)
    36852    0.048    0.000    0.233    0.000 context.py:2052(device)
   355356    0.392    0.000    0.725    0.000 context.py:215(push)
   355356    0.216    0.000    0.263    0.000 context.py:235(pop)
    12243    0.015    0.000    0.112    0.000 context.py:2400(async_wait)
    36759    0.039    0.000    1.458    0.000 context.py:2436(add_function)
    36774    0.027    0.000    0.201    0.000 context.py:2441(remove_function)
    12248    0.012    0.000    0.048    0.000 context.py:2480(_tmp_in_graph_mode)
   281938    0.072    0.000    0.072    0.000 context.py:503(ensure_initialized)
    12243    0.015    0.00

        1    0.000    0.000    0.000    0.000 def_function.py:509(__init__)
        1    0.000    0.000    0.000    0.000 def_function.py:512(__del__)
       14    0.000    0.000    0.001    0.000 def_function.py:534(__init__)
       14    0.000    0.000    0.000    0.000 def_function.py:606(_get_key_for_call_stats)
        2    0.000    0.000    0.000    0.000 def_function.py:629(_defun_with_scope)
        1    0.000    0.000    0.019    0.019 def_function.py:635(wrapped_fn)
        2    0.000    0.000    0.000    0.000 def_function.py:693(_defun)
        1    0.000    0.000    0.025    0.025 def_function.py:720(_initialize)
    24486    0.026    0.000    0.026    0.000 def_function.py:827(experimental_get_tracing_count)
    12243    0.004    0.000    0.004    0.000 def_function.py:860(_run_functions_eagerly)
    12243    0.163    0.000   21.696    0.002 def_function.py:864(__call__)
    12243    0.059    0.000   21.275    0.002 def_function.py:891(_call)
    12248    0.096    0.000  

       29    0.000    0.000    0.001    0.000 function_def_to_graph.py:76(<listcomp>)
       29    0.000    0.000    0.000    0.000 function_def_to_graph.py:81(<listcomp>)
       29    0.000    0.000    0.000    0.000 function_def_to_graph.py:85(<listcomp>)
       29    0.000    0.000    0.001    0.000 function_def_to_graph.py:88(<listcomp>)
       14    0.000    0.000    0.003    0.000 function_deserialization.py:122(_deserialize_function_spec_as_nonmethod)
        1    0.000    0.000    0.000    0.000 function_deserialization.py:163(setup_bare_concrete_function)
       13    0.000    0.000    0.001    0.000 function_deserialization.py:190(__init__)
       13    0.000    0.000    0.004    0.000 function_deserialization.py:225(recreate_function)
        1    0.004    0.004    0.161    0.161 function_deserialization.py:312(load_function_def_library)
       30    0.000    0.000    0.000    0.000 function_deserialization.py:330(<genexpr>)
      491    0.000    0.000    0.001    0.000 func

        7    0.000    0.000    0.000    0.000 load.py:1179(_maybe_add_serialized_attributes)
       14    0.000    0.000    0.000    0.000 load.py:1188(_get_keras_attr)
        1    0.000    0.000    0.248    0.248 load.py:124(__init__)
        1    0.000    0.000    0.000    0.000 load.py:128(<dictcomp>)
        1    0.000    0.000    0.000    0.000 load.py:171(_convert_node_paths_to_ints)
        1    0.000    0.000    0.001    0.001 load.py:193(_retrieve_all_filtered_nodes)
        6    0.000    0.000    0.000    0.000 load.py:218(_is_graph_network)
        1    0.000    0.000    0.000    0.000 load.py:248(__init__)
       32    0.000    0.000    0.000    0.000 load.py:250(_find_node_child)
        1    0.000    0.000    0.000    0.000 load.py:252(<dictcomp>)
        1    0.000    0.000    0.031    0.031 load.py:256(_load_all)
        1    0.000    0.000    0.000    0.000 load.py:266(del_tracking)
        1    0.000    0.000    0.002    0.002 load.py:269(_create_saveable_object_fact

   122965    0.022    0.000    0.022    0.000 object_identity.py:32(__init__)
    86884    0.030    0.000    0.030    0.000 object_identity.py:36(__init__)
    98113    0.026    0.000    0.035    0.000 object_identity.py:39(_assert_type)
    36875    0.009    0.000    0.009    0.000 object_identity.py:39(unwrapped)
      434    0.000    0.000    0.000    0.000 object_identity.py:43(_assert_type)
    98113    0.041    0.000    0.076    0.000 object_identity.py:51(__eq__)
      434    0.000    0.000    0.000    0.000 object_identity.py:55(__eq__)
   122965    0.035    0.000    0.053    0.000 object_identity.py:60(__hash__)
    86884    0.039    0.000    0.061    0.000 object_identity.py:64(__hash__)
      235    0.000    0.000    0.000    0.000 object_identity.py:78(__init__)
      123    0.000    0.000    0.000    0.000 object_identity.py:81(unwrapped)
   293904    0.340    0.000    0.915    0.000 op_callbacks.py:118(should_invoke_op_callbacks)
   416408    0.141    0.000    0.222    0.

       10    0.000    0.000    0.000    0.000 posixpath.py:71(join)
        2    0.000    0.000    0.000    0.000 py_checkpoint_reader.py:51(get_variable_to_dtype_map)
        2    0.000    0.000    0.000    0.000 py_checkpoint_reader.py:52(<dictcomp>)
        1    0.000    0.000    0.001    0.001 py_checkpoint_reader.py:66(get_tensor)
        2    0.001    0.001    0.001    0.001 py_checkpoint_reader.py:84(NewCheckpointReader)
        4    0.002    0.001    0.009    0.002 random_ops.py:215(random_uniform)
        4    0.000    0.000    0.000    0.000 random_seed.py:41(get_seed)
       23    0.000    0.000    0.000    0.000 re.py:188(match)
       29    0.000    0.000    0.000    0.000 re.py:198(search)
        1    0.000    0.000    0.000    0.000 re.py:250(compile)
       53    0.000    0.000    0.000    0.000 re.py:289(_compile)
   293869    0.164    0.000    0.212    0.000 registry.py:77(list)
   587738    0.418    0.000    0.960    0.000 registry.py:85(lookup)
       38    0.000  

       81    0.000    0.000    0.000    0.000 traceable_stack.py:35(set_filename_and_line_from_caller)
       81    0.000    0.000    0.000    0.000 traceable_stack.py:69(copy_metadata)
   147095    0.079    0.000    0.089    0.000 traceable_stack.py:77(__init__)
       81    0.000    0.000    0.000    0.000 traceable_stack.py:86(push_obj)
       58    0.000    0.000    0.000    0.000 tracking.py:127(_delete_tracking)
    36729    0.015    0.000    0.015    0.000 tracking.py:144(__init__)
    36729    0.018    0.000    0.018    0.000 tracking.py:147(resources)
    73458    0.066    0.000    0.074    0.000 tracking.py:155(resource_tracker_scope)
   171402    0.263    0.000    0.698    0.000 tracking.py:194(__init__)
   171452    0.026    0.000    0.026    0.000 tracking.py:218(_destroy_resource)
   171452    0.332    0.000    4.023    0.000 tracking.py:264(__del__)
  390/166    0.000    0.000    0.002    0.000 tracking.py:80(__setattr__)
        3    0.000    0.000    0.000    0.000 tra

   368025    0.281    0.000    0.281    0.000 {built-in method tensorflow.python._pywrap_tfe.TFE_Py_UID}
    24504    0.027    0.000    0.027    0.000 {built-in method tensorflow.python._pywrap_tfe.TFE_Py_VariableWatcherVariableAccessed}
        4    0.000    0.000    0.000    0.000 {built-in method tensorflow.python.lib.io._pywrap_file_io.FileExists}
        3    0.000    0.000    0.000    0.000 {built-in method tensorflow.python.lib.io._pywrap_file_io.Stat}
    36821    0.039    0.000    0.039    0.000 {built-in method tensorflow.python.platform._pywrap_tf2.is_enabled}
        1    0.001    0.001    0.001    0.001 {built-in method tensorflow.python.util._pywrap_checkpoint_reader.CheckpointReader_GetTensor}
    73458    0.138    0.000    0.138    0.000 {built-in method tensorflow.python.util._pywrap_utils.AssertSameStructureForData}
    73460    0.268    0.000    0.268    0.000 {built-in method tensorflow.python.util._pywrap_utils.AssertSameStructure}
   783557    0.874    0.000    0.

   12243    0.712    0.000    0.962    0.000 main.py:129(get_flipped_3d_array)
   
   So cumulatively less than a second was spent making the 3d arrays

208367  125.265    0.001  125.977    0.001 {built-in method tensorflow.python._pywrap_tfe.TFE_Py_FastPathExecute}

This seems like a culprit. I think the neural net is taking too long

28833/505    0.194    0.000  398.430    0.789 main.py:587(score)
    
Ah, so not much time is spent in score, but LOTS of time is spent in there cumulatively